**When should I not remove stop Words?** <br>
* during sentimental analysis
* while translating languagge
* chat bot question answere

In [1]:
import spacy
from spacy.lang.en.stop_words import STOP_WORDS


In [2]:
#we can see all stop words present in english language
STOP_WORDS

{"'d",
 "'ll",
 "'m",
 "'re",
 "'s",
 "'ve",
 'a',
 'about',
 'above',
 'across',
 'after',
 'afterwards',
 'again',
 'against',
 'all',
 'almost',
 'alone',
 'along',
 'already',
 'also',
 'although',
 'always',
 'am',
 'among',
 'amongst',
 'amount',
 'an',
 'and',
 'another',
 'any',
 'anyhow',
 'anyone',
 'anything',
 'anyway',
 'anywhere',
 'are',
 'around',
 'as',
 'at',
 'back',
 'be',
 'became',
 'because',
 'become',
 'becomes',
 'becoming',
 'been',
 'before',
 'beforehand',
 'behind',
 'being',
 'below',
 'beside',
 'besides',
 'between',
 'beyond',
 'both',
 'bottom',
 'but',
 'by',
 'ca',
 'call',
 'can',
 'cannot',
 'could',
 'did',
 'do',
 'does',
 'doing',
 'done',
 'down',
 'due',
 'during',
 'each',
 'eight',
 'either',
 'eleven',
 'else',
 'elsewhere',
 'empty',
 'enough',
 'even',
 'ever',
 'every',
 'everyone',
 'everything',
 'everywhere',
 'except',
 'few',
 'fifteen',
 'fifty',
 'first',
 'five',
 'for',
 'former',
 'formerly',
 'forty',
 'four',
 'from',
 'fron

In [3]:
len(STOP_WORDS)

326

In [4]:
nlp=spacy.load("en_core_web_sm") 

doc =nlp("we just opened our wings, the flyingpart is coming soon")

for token in doc:
    if token.is_stop:
        print(token)

we
just
our
the
is


In [25]:
# Function to remove stop words
def preprocess(text):
    doc=nlp(text)
    
    no_stop_words = [token.text for token in doc if not token.is_stop]
    return no_stop_words

In [6]:
preprocess("we just opened our wings, the flyingpart is coming soon")

['opened', 'wings', ',', 'flyingpart', 'coming', 'soon']

* function for Stop Words

In [7]:
# if we also want ot remove punctuation we can use .is_punct
#removing stop words
def preprocess(text):
    doc=nlp(text)
    
    no_stop_words = [token.text for token in doc if not token.is_stop and not token.is_punct]
    return " ".join(no_stop_words)
#using  " ".join we can convert list to the string

In [8]:
preprocess("we just opened our wings, the flyingpart is coming soon")
# here it has removed punctuation also which were present 

'opened wings flyingpart coming soon'

In [9]:
import pandas as pd

In [10]:
df=pd.read_json("doj_press.json",lines=True) # lines = True means new data on new line 
df.shape

(13087, 6)

In [11]:
df.head()

id                                              title  \
0     None       Convicted Bomb Plotter Sentenced to 30 Years   
1  12-919   $1 Million in Restitution Payments Announced t...   
2  11-1002  $1 Million Settlement Reached for Natural Reso...   
3   10-015  10 Las Vegas Men Indicted \r\nfor Falsifying V...   
4   18-898  $100 Million Settlement Will Speed Cleanup Wor...   

                                            contents  \
0  PORTLAND, Oregon. – Mohamed Osman Mohamud, 23,...   
1    WASHINGTON – North Carolina’s Waccamaw River...   
2        BOSTON– A $1-million settlement has been...   
3    WASHINGTON—A federal grand jury in Las Vegas...   
4  The U.S. Department of Justice, the U.S. Envir...   

                        date         topics  \
0  2014-10-01T00:00:00-04:00             []   
1  2012-07-25T00:00:00-04:00             []   
2  2011-08-03T00:00:00-04:00             []   
3  2010-01-08T00:00:00-05:00             []   
4  2018-07-09T00:00:00-04:00  [Environment]   

                                     components  
0            [National Security Division (NSD)]  
1  [Environment and Natural Resources Division]  
2  [Environment and Natural Resources Division]  
3  [Environment and Natural Resources Division]  
4  [Environment and Natural Resources Division]

In [12]:
df.describe

<bound method NDFrame.describe of             id                                              title  \
0         None       Convicted Bomb Plotter Sentenced to 30 Years   
1      12-919   $1 Million in Restitution Payments Announced t...   
2      11-1002  $1 Million Settlement Reached for Natural Reso...   
3       10-015  10 Las Vegas Men Indicted \r\nfor Falsifying V...   
4       18-898  $100 Million Settlement Will Speed Cleanup Wor...   
...        ...                                                ...   
13082   16-735  Yuengling to Upgrade Environmental Measures to...   
13083   10-473  Zarein Ahmedzay Pleads Guilty to Terror Violat...   
13084   17-045  Zimmer Biomet Holdings Inc. Agrees to Pay $17....   
13085   17-252  ZTE Corporation Agrees to Plead Guilty and Pay...   
13086   17-304  ZTE Corporation Pleads Guilty for  Violating U...   

                                                contents  \
0      PORTLAND, Oregon. – Mohamed Osman Mohamud, 23,...   
1        WASHINGT

In [13]:
df=df[df["topics"].str.len()!=0]
#here we are getting all the topics whose length is not 0 .so we are deleting topics with length 0 which means that they are empty

In [14]:
df.head()

id                                              title  \
4    18-898  $100 Million Settlement Will Speed Cleanup Wor...   
7   14-1412  14 Indicted in Connection with New England Com...   
19  17-1419  2017 Southeast Regional Animal Cruelty Prosecu...   
22  15-1562  21st Century Oncology to Pay $19.75 Million to...   
23  17-1404  21st Century Oncology to Pay $26 Million to Se...   

                                             contents  \
4   The U.S. Department of Justice, the U.S. Envir...   
7   A 131-count criminal indictment was unsealed t...   
19  The United States Attorney’s Office for the Mi...   
22  21st Century Oncology LLC, has agreed to pay $...   
23  21st Century Oncology Inc. and certain of its ...   

                         date                                 topics  \
4   2018-07-09T00:00:00-04:00                          [Environment]   
7   2014-12-17T00:00:00-05:00                  [Consumer Protection]   
19  2017-12-14T00:00:00-05:00                          [Environment]   
22  2015-12-18T00:00:00-05:00  [False Claims Act, Health Care Fraud]   
23  2017-12-12T00:00:00-05:00  [Health Care Fraud, False Claims Act]   

                                           components  
4        [Environment and Natural Resources Division]  
7                                    [Civil Division]  
19  [Environment and Natural Resources Division, U...  
22                                   [Civil Division]  
23           [Civil Division, USAO - Florida, Middle]

In [15]:
df.shape

(4688, 6)

In [16]:
df=df.head(100)# taking only 100 rows 

In [17]:
df.shape

(100, 6)

In [18]:
df["contents"].iloc[4]

'21st Century Oncology Inc. and certain of its subsidiaries and affiliates have agreed to pay $26 million to the government to resolve a self-disclosure relating to the submission of false attestations regarding the company’s use of electronic health records software and separate allegations that they violated the False Claims Act by submitting, or causing the submission of, claims for certain services provided pursuant to referrals from physicians with whom they had improper financial relationships. \xa0 “The Justice Department is committed to zealously investigating improper financial relationships that have the potential to compromise physicians’ medical judgment,” said Acting Assistant Attorney General Chad A. Readler of the Justice Department’s Civil Division.\xa0 “However, we will work with companies that accept responsibility for their past compliance failures and promptly take corrective action.”  \xa0 21st Century Oncology, which is headquartered in Fort Myers, Florida, owns a

In [19]:
len(df["contents"].iloc[4])

5504

In [20]:
df["contents_new"]=df["contents"].apply(preprocess)
#applying stop word to contents 

In [21]:
df.head(5)

id                                              title  \
4    18-898  $100 Million Settlement Will Speed Cleanup Wor...   
7   14-1412  14 Indicted in Connection with New England Com...   
19  17-1419  2017 Southeast Regional Animal Cruelty Prosecu...   
22  15-1562  21st Century Oncology to Pay $19.75 Million to...   
23  17-1404  21st Century Oncology to Pay $26 Million to Se...   

                                             contents  \
4   The U.S. Department of Justice, the U.S. Envir...   
7   A 131-count criminal indictment was unsealed t...   
19  The United States Attorney’s Office for the Mi...   
22  21st Century Oncology LLC, has agreed to pay $...   
23  21st Century Oncology Inc. and certain of its ...   

                         date                                 topics  \
4   2018-07-09T00:00:00-04:00                          [Environment]   
7   2014-12-17T00:00:00-05:00                  [Consumer Protection]   
19  2017-12-14T00:00:00-05:00                          [Environment]   
22  2015-12-18T00:00:00-05:00  [False Claims Act, Health Care Fraud]   
23  2017-12-12T00:00:00-05:00  [Health Care Fraud, False Claims Act]   

                                           components  \
4        [Environment and Natural Resources Division]   
7                                    [Civil Division]   
19  [Environment and Natural Resources Division, U...   
22                                   [Civil Division]   
23           [Civil Division, USAO - Florida, Middle]   

                                         contents_new  
4   U.S. Department Justice U.S. Environmental Pro...  
7   131 count criminal indictment unsealed today B...  
19  United States Attorney Office Middle District ...  
22  21st Century Oncology LLC agreed pay $ 19.75 m...  
23  21st Century Oncology Inc. certain subsidiarie...

In [22]:
df['contents_new'].iloc[4][:300]

'21st Century Oncology Inc. certain subsidiaries affiliates agreed pay $ 26 million government resolve self disclosure relating submission false attestations company use electronic health records software separate allegations violated False Claims Act submitting causing submission claims certain serv'

In [23]:
len(df['contents'].iloc[4])

5504

In [24]:
len(df['contents_new'].iloc[4])

4217

* you can see clearly deference that stop words has been removed from the data earlier thy were 5504 words now only 4217 remained